### PR TITLE
[staging] gnugrep: 3.7 -> 3.11

### LINKED ISSUE #228535: Update request: gnugrep 3.7 → 3.10
- Package name: gnugrep
- Latest released version: 3.10

- Current version on the unstable channel: 3.7
- Current version on the stable/release channel: 3.7

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

**Notify maintainers**

@edolstra @m00wl

Also, @trofi for the last version bump.

-----

Note for maintainers: Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by dasJ**
Also replace myself as a maintainer

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dasJ** [comment]
Draft for now because this breaks PCRE

**Comment by dasJ** [comment]
Should work now, @trofi can you verify that I didn't break the bootstrap tools?

**Comment by trofi** [comment]
At least on `linux` (and I think on `darwin` as well) we will need to add `pcre2` to allowed references (or substitute from `pcre`?). Otherwise `stdenv` fails to build as:

```
$ nix build --no-link -f. stdenv
error: output '/nix/store/jy1wr873b4ib5gplrg3118z8gjs2bpa1-stdenv-linux' is not allowed to refer to the following paths:
         /nix/store/grd3jawsc8xchi0zg92k91m1n5piasbj-pcre2-10.42
```

**Comment by dasJ** [comment]
Do I need them in `persistent` as well?

File: pkgs/stdenv/linux/bootstrap-tools-musl/scripts/unpack-bootstrap-tools.sh [hunk 1]
```diff
@@ -26,7 +26,7 @@ for i in $out/bin/* $out/libexec/gcc/*/*/*; do
         ./patchelf --set-interpreter $LD_BINARY --set-rpath $out/lib --force-rpath "$i"
 done
 
-for i in $out/lib/libpcre* $out/lib/libc.so; do
+for i in $out/lib/libpcre2* $out/lib/libc.so; do
```
Comment (trofi) [inline pkgs/stdenv/linux/bootstrap-tools-musl/scripts/unpack-bootstrap-tools.sh:29]: Similar to `glibc` case it has to handle our current `bootstrapTools` as well.

**Comment by dasJ** [comment]
I have successfully built `stdenv` from this branch now for x86_64-linux

**Review by trofi — APPROVED** [review]
LGTM. We can fix the

```
./pkgs/stdenv/linux/bootstrap-tools-musl/scripts/unpack-bootstrap-tools.sh
./pkgs/stdenv/linux/bootstrap-tools/scripts/unpack-bootstrap-tools.sh
```

a while after (hopefully before anyone tries to use newer `bootstrapTools`.

File: pkgs/tools/text/gnugrep/default.nix [hunk 2]
```diff
@@ -1,35 +1,34 @@
...
 
-  # https://git.savannah.gnu.org/cgit/gnulib.git/commit/?id=b50c6442e43d79471a31a2a202d3e50c0557446f
-  patches = lib.optional stdenv.hostPlatform.isLoongArch64 ./sigsegv-loongarch.patch;
+  patches = [
+    ./fix-dash-p-with-pcre2.patch
+  ];
 
-  # Perl is needed for testing
-  nativeBuildInputs = [ perl ] ++ lib.optional stdenv.hostPlatform.isLoongArch64 autoreconfHook;
+  checkInputs = [ perl ];
```
Comment (Artturin) [inline pkgs/tools/text/gnugrep/default.nix:23]:
```suggestion
  nativeCheckInputs = [ perl ];
```

**Comment by vcunat** [comment]
Today's cutoff for "breaking changes" in 23.05, so I wonder if this is worth squeezing in.  (I don't feel confident enough to merge now.)

**Comment by dasJ** [comment]
I personally wouldn't risk it, it's not like people have been anxiously waiting for this gnugrep update ;)

**Comment by dasJ** [comment]
@Artturin applied suggestion
@trofi can you look at my latest commit? It enables more tests but adds a dependency on `glibcLocales`. Is this wise? stdenv still builds for me with this applied.

**Comment by trofi** [comment]
`glibcLocales` expands to `null` for `musl` and other non-`glibc` targes:

```
$ nix build --no-link github:NixOS/nixpkgs/pull/228714/merge#pkgsMusl.stdenv -L
error:
       … while calling the 'derivationStrict' builtin

         at //builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'stdenv-linux'
         whose name attribute is located at /nix/store/55sz8bdibv3qq6saialdkkiijvsrrx0s-source/pkgs/stdenv/generic/default.nix:93:14

       … while evaluating attribute 'allowedRequisites' of derivation 'stdenv-linux'

         at /nix/store/55sz8bdibv3qq6saialdkkiijvsrrx0s-source/pkgs/stdenv/generic/default.nix:84:7:

           83|     lib.optionalAttrs (allowedRequisites != null) {
           84|       allowedRequisites = allowedRequisites
             |       ^
           85|         ++ defaultNativeBuildInputs ++ defaultBuildInputs;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: value is null while a set was expected
```

The simplest would be to probably avoid use of `.override()` and pick default locale set.

**Comment by dasJ** [comment]
Thanks for that input, this suggestion also adds chinese locales, bringing the skipped tests down to 9 (skipped by default because they are expensive).

The only things that could be worth fixing are:
- `pcre-ascii-digits: skipped test: PCRE 10.42 and older lack PCRE2_EXTRA_ASCII_BSD`
- `sjis-mb: skipped test: SJIS locale not found`

File: pkgs/tools/text/gnugrep/default.nix [hunk 3]
```diff
@@ -62,12 +61,14 @@ stdenv.mkDerivation {
     license = licenses.gpl3Plus;
 
     maintainers = [
-      maintainers.eelco
+      maintainers.das_j
```
Comment (m00wl) [inline pkgs/tools/text/gnugrep/default.nix:64]: fine for me.

**Review by m00wl — APPROVED** [review]
Thank you for your help on this!

I successfully built the new `gnugrep` and `stdenv` on x86_64 and aarch64 as well.

**Comment by dasJ** [comment]
Bumped to 3.11, this means I dropped the 3.10 regression fix.

**Review by pogoba — APPROVED** [review]

**Review by m00wl — APPROVED** [review]

**Comment by vcunat** [comment]
musl is in trouble, but [fortunately only](https://hydra.nixos.org/build/221700360) on aarch64:
  https://hydra.nixos.org/build/221679820

**Comment by risicle** [comment]
This appears to break `pkgsStatic.gnugrep` for me on x86_64 linux, with 4 tests failing: `test-nl_langinfo-mt`, `test-setlocale_null-mt-one`, `test-setlocale_null-mt-all`, `test-thread_create`.

**Comment by wegank** [comment]
So, add `!stdenv.hostPlatform.isMusl` to `doCheck`?

**Comment by vcunat** [comment]
Yes, I'd use that as the simplest workaround, in case nothing better is submitted.  It's surely better than to have no musl-based stdenv.  For example, we could conditionally drop just the tests failing now (a tiny minority).

**Comment by dasJ** [comment]
I'm currently investigating

**Comment by vcunat** [comment]
Hmm, on `x86_64-darwin` it always gets stuck during tests (four times at least now).  That's on Hydra: https://hydra.nixos.org/build/222612843#tabs-buildsteps
Locally it did pass for me on the first try &ndash; so maybe something about rosetta.  I don't know, but I expect we'd conditionally drop tests again?

**Comment by dasJ** [comment]
so `doCheck = !stdenv.hostPlatform.isDarwin`?

**Comment by wegank** [comment]
IIRC, @reckenrode also mentioned the x86_64-darwin issue, albeit for sandbox builds.

**Comment by reckenrode** [comment]
I’ve also encountered it trying to work on the Darwin stdenv bump on x86_64-darwin. It always hangs on my Apple Silicon MBP, but it builds on our Intel iMac. The MBP is running 13.4 while the iMac is still on 13.3.1(a). I don’t know if OS version makes a difference, or if it’s a Rosetta 2 issue the test is triggering.

**Comment by vcunat** [comment]
Apparently it succeeded in the end
```bash
nix build /nix/store/cprxh7g44qgvygmc2hv3l6xg3qkyl6sx-gnugrep-3.11 --dry-run
```
so I guess I'll let it be for now and reconsider `doCheck = !stdenv.hostPlatform.isDarwin` once it gets stuck again? (which I expect to happen in some future `staging-next`)

**Comment by reckenrode** [comment]
How long did it take to run? I left mine going for hours, and it never finished.

**Comment by vcunat** [comment]
I don't know how to find that.  The log is easy to get, as [the URL](https://hydra.nixos.org/log/43n62clwhbgwxqsg9lijza0d8vajr76x-gnugrep-3.11.drv) doesn't change on retries, but I can't see how to find which job built it (and see the time and machine).

**Comment by Artturin** [comment]
> I don't know how to find that. The log is easy to get, as [the URL](https://hydra.nixos.org/log/43n62clwhbgwxqsg9lijza0d8vajr76x-gnugrep-3.11.drv) doesn't change on retries, but I can't see how to find which job built it (and see the time and machine).

> checkPhase completed in 2 minutes 32 seconds

**Comment by reckenrode** [comment]
Are any of the builders actually x86_64-darwin hardware? It’s still stuck for me.

```
[1/158/203 built, 0.0 MiB DL] building gnugrep-3.11 (checkPhase): FAIL: test-c-stack.sh
```

It’s been stuck at that for hours. I even tried to rework my stdenv stuff to build gnugrep later (with clang 16), but it didn’t make a difference.

**Comment by vcunat** [comment]
I believe all are aarch64-darwin currently.  @cole-h should have a better grasp on this.

**Comment by cole-h** [comment]
ofborg has 3 x86 and 2 aarch64 macs, with the aarch64 macs enabling Rosetta and taking x86 builds whenever The Scheduler desires.

**Comment by vcunat** [comment]
@cole-h: this was too big for OfBorg, I meant Hydra where it did succeed after several attempts (but I don't know on which machine).

**Comment by cole-h** [comment]
According to https://hydra.nixos.org/machines, we only have 1 "real" x86_64-darwin machine (`destined-gannet`). (There's `stirred-oryx` too, but it's been dead basically since forever.)

**Comment by cole-h** [comment]
> Can anyone with Rosetta test this patch?
> 
> ```diff
> diff --git a/pkgs/tools/text/gnugrep/default.nix b/pkgs/tools/text/gnugrep/default.nix
> index 75c6f695073..2494ed1e49b 100644
> --- a/pkgs/tools/text/gnugrep/default.nix
> +++ b/pkgs/tools/text/gnugrep/default.nix
> @@ -17,7 +17,8 @@ stdenv.mkDerivation {
>    };
> 
>    # Some gnulib tests fail on Musl: https://github.com/NixOS/nixpkgs/pull/228714
> -  postPatch = if stdenv.hostPlatform.isMusl then ''
> +  # They also seem to fail on Rosetta: https://github.com/NixOS/nixpkgs/pull/228714#issuecomment-1574300624
> +  postPatch = if stdenv.hostPlatform.isMusl || stdenv.hostPlatform.system == "x86_64-darwin" then ''
>      sed -i 's:gnulib-tests::g' Makefile.in
>    '' else null;
> ```

Worked for me:

```
vin@catacendre nixpkgs % file result/bin/grep
result/bin/grep: Mach-O 64-bit executable x86_64
vin@catacendre nixpkgs % realpath ./result
/nix/store/vpcabk9qbxssqj3qcy9c3nag465mi3zw-gnugrep-3.11
```

**Pull request merged**